Add Shelly Pro 3EM sensors

thecode opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add support for Shelly Pro 3EM counters.
Notes:
- Device doesn't report energy counters so they will still be missing
- On my device power measured using incandescent light bulb is negative (also on device web UI) _This was discussed with Shelly HW engineers and is valid - reversing the clamp will create negative power factor which is OK if it is close to 1_

Picture showing we report identical data to the device data:
![image](https://user-images.githubusercontent.com/1858925/213942561-104aae1b-314a-468c-b970-e82a925241cb.png)


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes https://github.com/home-assistant/core/issues/85941
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

thecode committed:
Add Shelly Pro 3EM sensors

bieniu commented on homeassistant/components/shelly/sensor.py:
```diff
@@ -318,10 +319,78 @@ class RestSensorDescription(RestEntityDescription, SensorEntityDescription):
+        sub_key="a_aprt_power",
+        name="Phase A Apparent Power",
+        native_unit_of_measurement=UnitOfApparentPower.VOLT_AMPERE,
+        device_class=SensorDeviceClass.POWER,
```
```suggestion
        device_class=SensorDeviceClass.APPARENT_POWER,
```

bieniu commented on homeassistant/components/shelly/sensor.py:
```diff
@@ -318,10 +319,78 @@ class RestSensorDescription(RestEntityDescription, SensorEntityDescription):
+        sub_key="b_aprt_power",
+        name="Phase B Apparent Power",
+        native_unit_of_measurement=UnitOfApparentPower.VOLT_AMPERE,
+        device_class=SensorDeviceClass.POWER,
```
```suggestion
        device_class=SensorDeviceClass.APPARENT_POWER,
```

bieniu commented on homeassistant/components/shelly/sensor.py:
```diff
@@ -318,10 +319,78 @@ class RestSensorDescription(RestEntityDescription, SensorEntityDescription):
+        sub_key="c_aprt_power",
+        name="Phase C Apparent Power",
+        native_unit_of_measurement=UnitOfApparentPower.VOLT_AMPERE,
+        device_class=SensorDeviceClass.POWER,
```
```suggestion
        device_class=SensorDeviceClass.APPARENT_POWER,
```

bieniu reviewed (changes requested):
We should use `APPARENT_POWER` device class for apparent power sensors.

thecode committed:
Fix Apparent Power sensor device class

Co-authored-by: Maciej Bieniek <bieniu@users.noreply.github.com>

bieniu reviewed (commented):
We can also add `n_current` sensor:
```python
    "n_current": RpcSensorDescription(
        key="em",
        sub_key="n_current",
        name="N Current",
        native_unit_of_measurement=UnitOfElectricCurrent.AMPERE,
        device_class=SensorDeviceClass.CURRENT,
        state_class=SensorStateClass.MEASUREMENT,
        entity_registry_enabled_default=False,
    ),
```

bieniu commented on homeassistant/components/shelly/sensor.py:
```diff
@@ -318,10 +319,78 @@ class RestSensorDescription(RestEntityDescription, SensorEntityDescription):
+    "a_act_power": RpcSensorDescription(
+        key="em",
+        sub_key="a_act_power",
+        name="Phase A Active Power",
```
One more thing 😊 We should probably use the new entity naming style.
```suggestion
        name="Phase A active power",
```

thecode replied:
Yes, we should use the new style, fixed for this PR, we should make another PR to fix existing sensors

thecode commented:
> We can also add `n_current` sensor:
> 
> ```python
>     "n_current": RpcSensorDescription(
>         key="em",
>         sub_key="n_current",
>         name="N Current",
>         native_unit_of_measurement=UnitOfElectricCurrent.AMPERE,
>         device_class=SensorDeviceClass.CURRENT,
>         state_class=SensorStateClass.MEASUREMENT,
>         entity_registry_enabled_default=False,
>     ),
> ```

I added it and deleted it while testing, since currently this channel doesn't work (also in the device web UI) and I prefer not to add a sensor that is not working.

bieniu commented:
> since currently this channel doesn't work

If I understand correctly, for the N current value to appear, it is necessary to connect the N clamps. That's why I suggest that this entity be disabled by default.

thecode commented:
> > since currently this channel doesn't work
> 
> If I understand correctly, for the N current value to appear, it is necessary to connect the N clamp. That's why I suggest that this entity be disabled by default.

I connected a N clamp, I don't mind adding it, but I think we are just going to get issues about it not working.

bieniu commented:
> I connected a N clamp

If it doesn't work with the clamp connected, then there is no point in adding this entity. Thanks for the explanation.

thecode committed:
Adapt entity naming to new style

bieniu reviewed (approved):
LGTM 👍

bieniu commented:
> will this PR be added in 2023.1.8?

This is a new feature, so it needs to go through beta stage and will be available in 2023.2.
